pkg/networkservice/mechanisms/srv6/server.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) 2020 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package srv6
+
+import (
+	"context"
+	"math"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/ligato/vpp-agent/api/models/vpp"
+	vpp_l3 "github.com/ligato/vpp-agent/api/models/vpp/l3"
+	vpp_srv6 "github.com/ligato/vpp-agent/api/models/vpp/srv6"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/api/pkg/api/connection"
+	"github.com/networkservicemesh/api/pkg/api/connection/mechanisms/srv6"
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/sdk-vppagent/pkg/networkservice/vppagent"
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type srv6Server struct{}
+
+// NewServer provides a NetworkServiceServer chain elements that support the srv6 Mechanism
+func NewServer() networkservice.NetworkServiceServer {
+	return &srv6Server{}
+}
+
+func (v *srv6Server) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*connection.Connection, error) {
+	if err := v.appendInterfaceConfig(ctx, request.GetConnection(), true); err != nil {
+		return nil, err
+	}
+	return next.Server(ctx).Request(ctx, request)
+}
+
+func (v *srv6Server) Close(ctx context.Context, conn *connection.Connection) (*empty.Empty, error) {
+	if err := v.appendInterfaceConfig(ctx, conn, false); err != nil {
+		return nil, err
+	}
+	return next.Server(ctx).Close(ctx, conn)
+}
+
+func (v *srv6Server) appendInterfaceConfig(ctx context.Context, conn *connection.Connection, connect bool) error {
+	conf := vppagent.Config(ctx)
+	if mechanism := srv6.ToMechanism(conn.GetMechanism()); mechanism != nil {
```

Comment:
The nesting level of the whole function can be reduced by 'return nil' on reverted condition